python-checks/src/main/java/org/sonar/python/checks/NumpyListOverGeneratorCheck.java

```diff
@@ -0,0 +1,97 @@
+ */
+package org.sonar.python.checks;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.HasSymbol;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.cfg.fixpoint.ReachingDefinitionsAnalysis;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6714")
+public class NumpyListOverGeneratorCheck extends PythonSubscriptionCheck {
+
+  public static final String MESSAGE = "Pass a list to \"np.array\" instead of passing a generator.";
+  private ReachingDefinitionsAnalysis reachingDefinitionsAnalysis;
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> reachingDefinitionsAnalysis = new ReachingDefinitionsAnalysis((ctx.pythonFile())));
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, this::checkNumpyArrayCall);
+  }
+
+  private void checkNumpyArrayCall(SubscriptionContext ctx) {
+    CallExpression call = (CallExpression) ctx.syntaxNode();
+    Optional.ofNullable(call.calleeSymbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter("numpy.array"::equals)
+      .ifPresent(fqn -> checkGeneratorCallee(call, ctx));
+  }
+
+  private void checkGeneratorCallee(CallExpression call, SubscriptionContext ctx) {
+    List<Argument> argList = call.arguments();
+    if (argList.isEmpty()) {
+      return;
+    }
+
+    if (Optional.of(argList.get(0))
+      .filter(arg -> arg.is(Tree.Kind.REGULAR_ARGUMENT))
+      .map(RegularArgument.class::cast)
+      .map(RegularArgument::expression)
+      .filter(regArg -> (regArg.is(Tree.Kind.GENERATOR_EXPR) || this.isNamedGeneratorExpression(regArg)))
+      .isEmpty()) {
+      return;
+    }
+
+    if (Optional.ofNullable(TreeUtils.nthArgumentOrKeyword(1, "dtype", argList))
+      .filter(regArg -> regArg.expression().is(Tree.Kind.NAME))
+      .map(regArg -> (Name) regArg.expression())
+      .map(HasSymbol::symbol)
+      .map(Symbol::fullyQualifiedName)
+      .filter("object"::equals)
+      .isEmpty()) {
+      ctx.addIssue(call, MESSAGE);
+    }
+  }
+
+  private boolean isNamedGeneratorExpression(Expression expression) {
+    return Optional.of(expression)
+      .filter(expr -> expr.is(Tree.Kind.NAME))
+      .map(Name.class::cast)
+      .map(name -> this.reachingDefinitionsAnalysis.valuesAtLocation(name))
+      .filter(this::checkSetProperties)
+      .isPresent();
+  }
+
+  private boolean checkSetProperties(Set<Expression> set) {
```

Comment:
Here we could change the method to a static one. SonarQube will detect this issue, you can check the code smells on PR.

Reply:
Corrected in 0f9c655b3d898b729177f761f78d055a58a8f31d.